File: Orbit/Scala/Orbit/src/main/scala/orbit/Orbit.scala
package orbit
import benchmark._
import scala.annotation.tailrec
import scala.util.Try
import orbit.util.{ ScalaSets, JavaSets }

object Main extends App {
  def run[A](body: => (collection.Set[A], Long)) {
    val (res, time) = body
    println(s"Result: ${res.size} - Time Elapsed: $time ms")
  }

  def solvers(sets: util.SetProvider, M: Int, G: Int) = {
    val simple = new solver.Simple(sets)
    val akka = new solver.Akka(sets)
    val akkaSystem = new solver.AkkaSystem(sets)
    type D = Definition
    Map(
      (1, ("Sequential with Immutable Set", 					(x: D) => simple.solve(x))),
      (2, ("Sequential with Mutable Set", 						(x: D) => simple.solve2(x))),
      (3, ("Parallel Collections", 								(x: D) => simple.solveParSeq(x, M))),
      (4, ("Parallel Collections with Concurrent Map", 			(x: D) => simple.solveParSeqWithConcurrentMap(x, M))),
      (5, ("Futures", 											(x: D) => simple.solveFuture(x, M, G))),
      (6, ("Akka with Immutable Set", 							(x: D) => akka.solveImmutableSet(x, G))),
      (7, ("Akka with Mutable Set", 							(x: D) => akka.solveMutableSet(x, G))),
      (8, ("Akka with Concurrent Map", 							(x: D) => akka.solveConcurrentMap(x, G))),
      (9, ("Akka System with Actor Workers", 					(x: D) => akkaSystem.solveActorWorkers(x, M, G))),
      (10, ("Akka System with Actor Workers & Concurrent Map", 	(x: D) => akkaSystem.solveActorWorkersConcurrentMap(x, M, G))))
  }

  print("nOfTimes each test will run (default = 10): ")
  val times = Try(readLine().toInt) getOrElse 10
  print("Choose mode [1 -> int64, 2 -> bigint] (default = int64): ")
  val mode = Try(readLine().toInt) getOrElse 1
  print("Give me nOfMappers (default = ProcessorCount): ")
  val M = Try(readLine().toInt) getOrElse Runtime.getRuntime().availableProcessors()
  print("Give me chunkSize (default = 1000): ")
  val G = Try(readLine().toInt) getOrElse 1000
  print("Set Implementation [1 -> Scala, 2 -> Java] (default = 1): ")
  val sets = readLine() match {
    case "2" => JavaSets
    case "1" | _ => ScalaSets
  }
  print("""Choose Implementation from [
    1 -> Sequential with Immutable Set, 
    2 -> Sequential with Mutable Set,
    3 -> Parallel Collections,
    4 -> Parallel Collections with Concurrent Map,
    5 -> Futures,
    6 -> Akka with Immutable Set,
    7 -> Akka with Mutable Set,
    8 -> Akka with Concurrent Map, 
    9 -> Akka System with Actor Workers,
    10 -> Akka System with Actor Workers & Concurrent Map
] (default = 1): """)
  val implementation = Try(readLine().toInt).getOrElse(1)
  print("Give me l,d,f (space separated on the same line, f <= 10, default = 200000 10000 8): ")
  val Array(l, d, f) = Try {
    readLine() split (" ", 3) map (_.trim.toInt)
  }.getOrElse(Array(200000, 10000, 8))

  println()

  val problem = mode match {
    case 1 =>
      new GenBench[Long](l, d, f)
    case 2 =>
      new GenBench[BigInt](l, d, f)
    case _ =>
      throw new Error("No such option available")
  }

  val solve = solvers(sets, M, G)(implementation)._2
  for (i <- 1 to times)
    run { solve(problem) }
}

File: Orbit/Erlang/orbit.erl
-module(orbit).
-include("definition.hrl").
-export([repeat/2, timedRun/3, run/0, main/1]).

% Timed Run
timedRun(Fun, Args) ->
  {T, Xs} = timer:tc(Fun, [Args]),
  {length(Xs), T div 1000}.

timedRun(Module, Fun, Args) ->
  {T, Xs} = timer:tc(Module, Fun, [Args]),
  {length(Xs), T div 1000}.

toIntegerOrElse(Str, Default) ->
  case string:to_integer(Str) of
    {error,_} -> Default;
    {N, _} -> N
  end.

solveFuncs(M, G) ->
	dict:from_list([
		{1, fun(X) -> solve:solve(X) end},
		%{2, fun(X) -> solve:solve_conc(X, G, fun solve:solve_conc_helper/4) end}, 
		{2, fun(X) -> solve:solve_conc(X, G, fun solve:solve_conc_helper2/4) end},
		{3, fun(X) -> solve:solve_conc(X, G, fun solve:solve_conc_helper3/4) end},
    {4 ,fun(X) -> solve:solve_conc_with_workers(X, M, G) end}
	]).

repeat(Body, 1) -> 
  Body();
repeat(Body, N)->
  Body(),
  repeat(Body, N-1).

run() ->
  Times = toIntegerOrElse(io:get_line('nOfTimes each test will run (default = 10): '),10), io:nl(),
  M = toIntegerOrElse(io:get_line('Give me nOfMappers (default = schedulers_online): '), erlang:system_info(schedulers_online)),io:nl(),
  G = toIntegerOrElse(io:get_line('Give me chunkSize (default = 1000): '), 1000),io:nl(),
  io:fwrite("Choose Implementation from [\n"),
  io:fwrite("    1 -> Sequential,\n"),
  io:fwrite("    2 -> Concurrent Tasks with Sequential Ets,\n"),
  io:fwrite("    3 -> Concurrent Tasks with Concurrent Ets \n"),
  io:fwrite("    4 -> Concurrent with Workers and Concurrent Ets\n"),
  Implementation = toIntegerOrElse(io:get_line("] (default = 1): "),1),
  Map = solveFuncs(M, G),
  L = toIntegerOrElse(io:get_line("L (default = 200000): "),200000),io:nl(),
  D = toIntegerOrElse(io:get_line("D (default = 10000): "),10000),io:nl(),
  F = toIntegerOrElse(io:get_line("F (default = 8): "),8),io:nl(),
  io:format("Schedulers = ~p, Times = ~p, M = ~p, G = ~p, Implementation = ~p, L = ~p, D = ~p, F = ~p", [erlang:system_info(schedulers_online),Times, M, G, Implementation, L, D, F]),
  repeat(fun() ->
    {Res, T} = timedRun(dict:fetch(Implementation, Map), benchmarks:simple(L,D,F)),
    io:format("\nResult: ~p - Time Elapsed: ~p ms",[Res,T]) 
  end, Times).


main(_Args) -> run().